File: General/extract_fasta_entries.pl
#!/usr/bin/env perl
"""
Retrieves fasta sequences from fasta master file using names from a text file. 

Usage: extract_fasta_entries.pl <list_of_names.txt> <master_sequences.fasta> <output_sequences.fasta>

Updated 180721 by Ronald Cutler
"""
my $list_file = $ARGV[0];
my $fasta_in = $ARGV[1];
my $fasta_out = $ARGV[2];
open(LIST_FILE, "<", $list_file) or die "could not open '$list_file' : $! \n";
open(FASTA_IN, "<", $fasta_in) or die "could not open '$fasta_in' : $! \n";
open(FASTA_OUT, ">", $fasta_out) or die "could not open $fasta_out : $! \n";
my @headers = ();
while(<LIST_FILE>) {
    chomp;
    next if ( /^\s*$/ );
    push(@headers, $_);
}
my $pat = join '|', map quotemeta, @headers;
$/ = ">";
while(<FASTA_IN>) {
    chomp;
    if ( /$pat/ ) { print FASTA_OUT ">$_"; }
}
close(LIST_FILE);
close(FASTA_IN);
close(FASTA_OUT);

File: General/extract_fasta_entries.py
#!/usr/bin/env python
"""
For reading in a txt file of gene names on each line and a multi-fasta
database which contains the sequence associated with these name and 
returns a multi-fasta file of fasta entries corresponding to the input. 

USAGE: Extract_cDNA.py <geneList.txt> <multi_fasta.fasta> <output.txt>

Last updated 170811 by Ronald Cutler

TODO: Use fuzzy matching when names are not 1-1 match.
"""

import sys

def main(query, ref, output):

	geneList = []
	with open( query, 'r') as reader1:
		for line in reader1:
			geneList.append(line)

	hitList = []
	with open(ref, 'r') as reader2:
		for line in (reader2):
			if ">" in line:
				if line[1:] in geneList:
					hit = line + next(reader2) + '\n'
					hitList.append(hit)

	with open(output, 'w') as writer:
		for item in hitList:
			writer.write(item)



if __name__ == "__main__":
	main(sys.argv[1], sys.argv[2], sys.argv[3])